docs/source/optimising.rst

```diff
@@ -0,0 +1,198 @@
+
+   * Upload the file to `speedscope <https://www.speedscope.app/>`_ and view it there.
+
+Adding your own events
+~~~~~~~~~~~~~~~~~~~~~~
+
+It is very easy to add your own events to the flame graph and there
+are a few different ways of doing it. The simplest methods are:
+
+* With a context manager:
+
+  .. code-block:: python
+      
+      from firedrake.petsc import PETSc
+
+      with PETSc.Log.Event("foo"):
+          do_something_expensive()
+
+* With a decorator:
+
+  .. code-block:: python
+
+      from firedrake.petsc import PETSc
+
+      @PETSc.Log.EventDecorator("foo")
+      def do_something_expensive():
+          ...
+
+  If no arguments are passed to ``PETSc.Log.EventDecorator`` then the
+  event name will be the same as the function.
+
+Caveats
+~~~~~~~
+
+* The ``flamegraph.pl`` script assumes by default that the values
+  in the stack traces are sample counts. This means that if you
+  hover over functions in the SVG it will report the count in terms
+  of 'samples' rather than the correct unit of microseconds. A simple
+  fix to this is to include the command line option ``--countname us``
+  when you generate the SVG. For example:
+
+  .. code-block:: bash
+
+    $ ./flamegraph.pl --countname us foo.txt > foo.svg
+
+* If you use PETSc stages in your code these will be ignored in the flame graph.
+
+* If you call ``PETSc.Log.begin()`` as part of your script/package
+  then profiling will not work as expected. This is because this
+  function starts PETSc's default (flat) logging while we need to
+  use nested logging instead.
+
+  This issue can be avoided with the simple guard:
+
+  .. code-block:: python
+  
+    from firedrake.petsc import OptionsManager
+
+    # If the -log_view flag is passed you don't need to call 
+    # PETSc.Log.begin because it is done automatically.
+    if "log_view" in OptionsManager.commandline_options:
+        PETSc.Log.begin()
+
+Common performance issues
+-------------------------
+
+Calling ``solve`` repeatedly
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When solving PDEs, Firedrake uses a PETSc ``SNES`` (nonlinear solver)
+under the hood. Every time the user calls :py:func:`~firedrake.solving.solve`
+a new ``SNES`` is created and used to solve the problem. This is a
+convenient shorthand for scripts that only need to solve a problem
+once, but it is fairly expensive to set up a new ``SNES`` and so
+repeated calls to :py:func:`~firedrake.solving.solve` will introduce
+some overhead.
+
+To get around this problem, users should instead instantiate
+a variational problem (e.g. :py:class:`~.LinearVariationalProblem`)
+and solver (e.g. :py:class:`~.LinearVariationalSolver`) outside of
+the loop body. An example showing how this is done can be found
+in `this demo <https://firedrakeproject.org/demos/DG_advection.py.html>`_.
+
+Other useful tools
+------------------
+
+Here we present a handful of performance analysis tools that users may
+find useful to run with their codes.
+
```

Comment:
pyinstrument?

Reply:
I haven't used this until now. Neat tool! Will add.